Einsum ellipsis

fukatani opened the pull request:
Merge #404 first, please.

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -64,22 +78,49 @@ def calc_summed_view(ioperand, input_subscript, output_subscript):
+        raise ValueError('`source` and `destination` arguments must have '
+                         'the same number of elements')
+
+    order = [n for n in range(a.ndim) if n not in source]
```
six.moves.range

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -18,22 +18,33 @@ def calc_single_view(ioperand, subscript):
     """
 
-    assert ioperand.ndim == len(subscript)
+    if '@' not in subscript:
```
`assert ioperand.ndim == len(subscript) - subscript.count('@')`

fukatani replied:
Because `@` and `...` can contain 0 or more dimension, `ioperand.ndim == len(subscript) - subscript.count('@')` is not always `True`.
So I added assertion l.21- l.24.

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -18,22 +18,33 @@ def calc_single_view(ioperand, subscript):
+                    axes_to_diag.append(i)
+                else:
+                    axes_to_diag.append(i - len(subscripts_excluded_at))
+        axes_to_diag = numpy.core.numeric.normalize_axis_tuple(axes_to_diag,
```
`normalize_axis_tuple` was added v 1.13.0. We need to support older versions. Please reimplement it.

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -64,22 +78,49 @@ def calc_summed_view(ioperand, input_subscript, output_subscript):
+
+    order = [n for n in range(a.ndim) if n not in source]
+
+    for dest, src in sorted(zip(destination, source)):
```
six.moves.zip

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -96,13 +137,43 @@ def calc_transposed_view(ioperand, input_subscript, output_subscript):
+def move_broadcast_axes_to_front(ioperands, subscripts):
+    broadcasted_operands = []
+    broadcasted_subscripts = []
+    for operand, subscript in zip(ioperands, subscripts):
```
six.move.zip

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -96,13 +137,43 @@ def calc_transposed_view(ioperand, input_subscript, output_subscript):
+    for operand, subscript in zip(ioperands, subscripts):
+        if '@' in subscript:
+            ellipsis_pos = subscript.find('@')
+            source_axes = list(range(ellipsis_pos))
```
six.moves.range

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -96,13 +137,43 @@ def calc_transposed_view(ioperand, input_subscript, output_subscript):
+        if '@' in subscript:
+            ellipsis_pos = subscript.find('@')
+            source_axes = list(range(ellipsis_pos))
+            destination_axes = [i - len(source_axes) for i in source_axes]
```
`len(source) == ellipsis_pos` ?

fukatani replied:
Thanks.

I refactored to
```
destination_axes = [i - ellipsis_pos for i in source_axes]
```
.

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -112,19 +183,46 @@ def calc_combined_view(ioperands, subscripts):
+    a_shape_stack = []
+    b_shape_stack = []
+    is_first_operand = True
+    for operand, subscript in zip(ioperands, subscripts):
```
six.moves.zip

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -112,19 +183,46 @@ def calc_combined_view(ioperands, subscripts):
+        if subscript and '@' == subscript[0]:
+            broadcasted_dims = operand.ndim - len(subscript) + 1
+            a_shape = numpy.prod(operand.shape[:broadcasted_dims],
+                                 dtype=numpy.int8)
```
Why do you use int8? Is it safe?

fukatani replied:
Changed to `uint8`.

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -112,19 +183,46 @@ def calc_combined_view(ioperands, subscripts):
+            if len(operand.shape[:broadcasted_dims]) > len(a_shape_stack):
+                a_shape_stack = list(operand.shape[:broadcasted_dims])
+            b_shape = numpy.prod(operand.shape[broadcasted_dims:],
+                                 dtype=numpy.int8)
```
ditto

fukatani commented on cupy/linalg/einsum.py:
```diff
@@ -209,10 +301,17 @@ def einsum(*operands):
                 continue
             raise ValueError('einstein sum subscripts string includes output '
                              'subscript \'{}\' multiple times'.format(key))
+        if '@' in output_subscript and re.match('/^[a-zA-Z]*@[a-zA-Z]*?$/',
```
I thought again, I have to check only `output_subscript.count('@') >= 2`.

fukatani commented on cupy/linalg/einsum.py:
```diff
@@ -209,10 +301,17 @@ def einsum(*operands):
+                                                output_subscript):
+            raise ValueError('Two or more \'...\' ellipsis can\'t be used for'
+                             'output subscript')
+        if '@' in input_subscripts and '@' not in output_subscript:
```
Thanks!
I researched `numpy.einsum` implementation and behavior.
As a result, if '...' does not point any axis, subscript such as `i...->i` is permitted.
But if '...' points any axis in input, output must contain `...`.

ex.
```
A = numpy.arange(8).reshape(2, 4)

# OK
B = numpy.einsum('ij...->ij', A)
B = numpy.einsum('i...j->ij', A)
B = numpy.einsum('...ij->ij', A)

# raise error
B = numpy.einsum('ij->i...', A)
B = numpy.einsum('ij->...j', A)
```

fukatani replied:
And I found, numpy.einsum('ij->...ij', A) is allowed.
I added testcases.
```
B = numpy.einsum('ij->...ij', A)
B = numpy.einsum('ij->i...j', A)
B = numpy.einsum('ij->ij...', A)
```

fukatani commented:
Thank you for your attentive review.
I addressed your comments.

And I can't understand why travis CI failed with Python 3.5.
I appreciate if you inspect travis error.

unnonouno commented:
Thank you for quick fix. There is another problem for Py3.5. Please ignore it.

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -187,7 +280,12 @@ def einsum(*operands):
             converted_inputs.append(cupy.asarray(a, dtype=dtype))
 
-    match = re.match('^([a-zA-Z,]+)(->[a-zA-Z]*)?$', subscripts)
+    subscripts = subscripts.replace('...', '@')
```
`...` is replaced with `@`  for simplicity of implementation, right? If so please write a comment about it.

fukatani replied:
Yes, I should add comment.

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -64,22 +77,57 @@ def calc_summed_view(ioperand, input_subscript, output_subscript):
+    return tuple((ax % ndim) for ax in axis)
+
+
+# TODO(fukatani): Implement as cupy.moveaxis
```
Do you have a plant to make `moveaxis`?

fukatani replied:
Yes.

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -64,22 +77,57 @@ def calc_summed_view(ioperand, input_subscript, output_subscript):
+
+    .. seealso:: :func:`numpy.core.numeric.normalize_axis_tuple`
+    """
+    return tuple((ax % ndim) for ax in axis)
```
Is it safe when `ax` is larger than `2 * ndim`?

fukatani replied:
I added exception if ax > ndim as `numpy.normalize_axis_tuple`.

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -18,22 +18,32 @@ def calc_single_view(ioperand, subscript):
     """
 
-    assert ioperand.ndim == len(subscript)
+    if '@' in subscript:
```
I think at most one `@` exist. If so, when `'@' in subscript` `subsceript.count('@') == 1`. Is it right?

fukatani replied:
Ah, Yes.
I make assertion more simple. (l.22-l.23.)

fukatani commented:
I'm also studying einsum, anyway thanks for your review.
I addressed your comments.

unnonouno commented:
jenkins test this please

unnonouno commented:
In my environment, this error occurs:
```
AssertionError: The state of c_contiguous flag is false. (cupy_result:False numpy_result:True)
```

fukatani commented:
Sorry for error!
But I haven't seen such a error, and all tests have passed in my environment.
Could you give me more information about error?

unnonouno commented:
For example like this:
```
======================================================================
FAIL: test_einsum_binary (cupy_tests.linalg_tests.test_einsum.TestEinSumBinaryOperation_param_9)  parameter: {'shape_a': (2, 3, 4), 'subscripts': 'ij...,j...->ij...', 'shape_b': (3,), 'skip_overflow': False}
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/unno/git/cupy/cupy/testing/helper.py", line 737, in test_func
    impl(self, *args, **kw_copy)
  File "/home/unno/git/cupy/cupy/testing/helper.py", line 125, in test_func
    check_func(cupy_result, numpy_result)
  File "/home/unno/git/cupy/cupy/testing/helper.py", line 189, in check_func
    c.flags.c_contiguous, n.flags.c_contiguous))
AssertionError: The state of c_contiguous flag is false. (cupy_result:False numpy_result:True)
-------------------- >> begin captured stdout << ---------------------
{'dtype_b': <type 'numpy.int8'>, 'dtype_a': <type 'numpy.float32'>}

--------------------- >> end captured stdout << ----------------------
```

unnonouno commented:
jenkins test this please

fukatani commented:
I understood why test is failure.
`cupy.transpose` changes array contiguity, but if order argument is 'K', numpy.einsum keeps array contiguity.

```
order : {‘C’, ‘F’, ‘A’, ‘K’}, optional
Controls the memory layout of the output. ‘C’ means it should be C contiguous. ‘F’ means it should be Fortran contiguous, ‘A’ means it should be ‘F’ if the inputs are all ‘F’, ‘C’ otherwise. ‘K’ means it should be as close to the layout as the inputs as is possible, including arbitrarily permuted axes. Default is ‘K’.
```

I'm investigating what "‘K’ means it should be as close to the layout as the inputs as is possible" means.

fukatani commented:
@unnonouno 
Sorry, I want to consult with you about contiguity.

In conclusion, it is difficult for me to keep contiguity compatibility for all cases.

`numpy.einsum` result contiguity depends on the course if order='K', as I investigated.

(Please see npyiter_apply_forced_iteration_order in https://github.com/numpy/numpy/blob/d55f40b1e48d5a5fbed80e00a140c9db6e19732f/numpy/core/src/multiarray/nditer_constr.c)

/////////

How about add the notation "cupy.einsum may break contiguouity with `numpy.einsum` about result contiguouity." and not to manipulate contiguity explicitly in this PR?

I think `einsum` users who care about contiguity should use order='C' or 'F' or 'A'.
These options are not supported in `cupy` currently, but they are easy to implement.

unnonouno commented:
'K' contiguity is very difficult (its specification is not documented enough....). I think it is OK not to support all contiguous option.

fukatani commented:
OK!

For now, I fixed test to ignore contiguity.
Default option is 'K'.
When 'C', 'F', and 'A' are supported in cupy in future, memory contiguity should be compatible with numpy in these option, I think.

unnonouno commented:
Thank you! Jenkins test this please

unnonouno commented on cupy/linalg/einsum.py:
```diff
@@ -133,24 +241,20 @@ def einsum(*operands):
     """einsum(subscripts, *operands)
 
     Evaluates the Einstein summation convention on the operands.
-
```
These empty lines are required. Check sphinx result.

unnonouno commented:
Sphinx failed to make a document. Please check the last comment.

fukatani commented:
Unfortunately, I can't build cupy doc.
I installed sphinx, install cupy by `python setup.py install` and `make html` in doc.
but,
```
The 'cupy' distribution was not found and is required by the application
```

So, I did fix doc in blind.

unnonouno commented:
OK. I'll check it.
Jeknins test this please.

unnonouno commented:
LGTM
